### PR TITLE
[Card#23]-Admin dashboard-Dependency on display_leaves_on_dashboard branch

### PR DESCRIPTION
**Pull request by HarshwardhanSingh**
Trello card at https://trello.com/c/YvH7gDnL & https://trello.com/c/kmTj8z3p

Branched out from **master**

Dependent on **display_leaves_on_dashboard** branch

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -14,4 +14,41 @@
...
+$(function () {
+    $('.start').datetimepicker({
+      format: 'DD/MM/YYYY'
+    });
+    $('.end').datetimepicker({
+        useCurrent: false,
+        format: 'DD/MM/YYYY'
+    });
+    $(".start").on("dp.change", function (e) {
+        $('.end').data("DateTimePicker").minDate(e.date);
```
Comment (deepak) [inline app/assets/javascripts/application.js]: why is this needed ?

File: app/views/leaves/_form.html.erb [hunk 2]
```diff
@@ -1,15 +1,51 @@
-<%= form_for(calendar_leave_index_path(params[:calendar_id]), method: "post", class: "js-leave-form", url: calendar_leave_index_path(params[:calendar_id])) do%>
+<%= form_for(create_leave_path(@leave), method: "post", class: "js-leave-form", url: create_leave_path(@leave)) do%>
   <div class="form-group">
-    <%= text_field_tag("summary", nil,class: "form-control", placeholder: "Summery") %>
+    <%= collection_select(:leave, :reason_id, Reason.all, :id, :name) %>
```
Comment (deepak) [inline app/views/leaves/_form.html.erb]: views should not do database queries. make sure you pass `reasons` as a locals like any other, while rendering the view

File: app/views/pages/admin_dashboard.html.erb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    <h4><%= current_user.name %></h4>
+    <h5>ADMINISTRATOR</h5>
+  </div>
+  <div class="col-md-8 left-bar">
+    <h4>Average Leave Taken from all employees</h4>
+    <%= render partial: 'partials/filter_form' %>
+    <br/>
+  </br/>
+    <table class="table" id="table">
+      <% Reason.all.each do |reason| %>
```
Comment (deepak) [inline app/views/pages/admin_dashboard.html.erb]: views should not do database queries. make sure you pass `reasons` as a locals like any other, while rendering the view.

also the table is queried twice in this view

File: app/views/pages/filter_leaves.js.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+$("table#table").html("<%= escape_javascript render(partial: 'partials/table', locals:{leaves: @leaves}) %>");
```
Comment (deepak) [inline app/views/pages/filter_leaves.js.erb]: table and then again id of table! why ? 
better to have something more descriptive id than table 

also the id needs to be prefixed with `js-` 
and I do not mean that it should be named `js-table`

File: app/views/pages/filter_leaves_users.js.erb [hunk 5]
```diff
@@ -0,0 +1 @@
+$("table#user").html("<%= escape_javascript render(partial: 'partials/table', locals:{leaves: @leaves}) %>");
```
Comment (deepak) [inline app/views/pages/filter_leaves_users.js.erb]: same as before. https://github.com/code-mancers/leave-tracker/pull/25/files#r72599102

File: app/views/partials/_table.html.erb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<table class="table">
+  <% Reason.all.each do |reason| %>
```
Comment (deepak) [inline app/views/partials/_table.html.erb]: view should not query data. should be passed in like any other locals

File: db/schema.rb [hunk 7]
```diff
@@ -10,49 +10,70 @@
...
-    t.datetime "end"
-    t.integer  "user_id"
+    t.string   "summary",                null: false
+    t.datetime "start",                  null: false
+    t.datetime "end",                    null: false
+    t.datetime "created_at",             null: false
+    t.datetime "updated_at",             null: false
+    t.integer  "user_id",                null: false
+    t.integer  "reason_id",              null: false
+    t.integer  "leave_type", default: 0, null: false
```
Comment (deepak) [inline db/schema.rb]: does the UI default to enum value 0 ? 
even if yes. better not be have a database default and couple the UI and database in this way

File: spec/models/reason_spec.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (deepak) [inline spec/models/reason_spec.rb]: auto-generated tests with no actual test, can be removed

**Comment by deepak** [comment]
@HarshwardhanSingh have some comments. please merge once comments are fixed

File: db/schema.rb (hunk 7 above)
Comment (deepak) [inline db/schema.rb]: edit: after discussion. not a big deal

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (HarshwardhanSingh) [inline app/assets/javascripts/application.js]: @deepak if you asked about line 46 then is is required to prevent user from passing end date less than start date.

File: app/views/pages/admin_dashboard.html.erb (hunk 3 above)
Comment (HarshwardhanSingh) [inline app/views/pages/admin_dashboard.html.erb]: @deepak  Fixed. Now passing reasons via locals. As for the table queried twice that's because one table returns avg data of all employees and 2nd table filters leaves for selected user.

Basically both tables performs different operations and hence required.

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (deepak) [inline app/assets/javascripts/application.js]: ok
